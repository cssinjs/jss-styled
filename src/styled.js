import {PureComponent, createElement} from 'react'
import {getDynamicStyles} from 'jss'

import filterProps from './utils/filterProps'
import composeClasses from './utils/composeClasses'
import generateTagName from './utils/generateTagName'

import type {
  JssStaticSheet,
  JssDynamicSheet,
  ComponentStyleType,
  StyledElementPropsType
} from './types'

type StyledArgs = {
  tagName: string,
  elementStyle: ComponentStyleType,
  mountSheets: Function,
  addRule: Function
}

const styled = ({tagName, elementStyle, mountSheets, addRule}: StyledArgs) => {
  const dynamicStyle = getDynamicStyles(elementStyle)
  const staticTagName = generateTagName(tagName)

  return class StyledElement extends PureComponent {
    static tagName: string = tagName
    static style: ComponentStyleType = elementStyle

    props: StyledElementPropsType

    dynamicTagName = ''
    staticSheet: JssStaticSheet
    dynamicSheet: JssDynamicSheet

    constructor(props: StyledElementPropsType) {
      super(props)
      if (!this.dynamicTagName) {
        this.dynamicTagName = generateTagName(tagName)
      }
    }

    componentWillMount() {
      Object.assign(this, mountSheets())

      if (!this.staticSheet.getRule(staticTagName)) {
        addRule(staticTagName, elementStyle)
      }

      if (dynamicStyle && !this.dynamicSheet.getRule(this.dynamicTagName)) {
        addRule(this.dynamicTagName, dynamicStyle, this.props)
      }
    }

    componentWillReceiveProps(nextProps: StyledElementPropsType) {
      if (dynamicStyle) {
        this.dynamicSheet.update(this.dynamicTagName, nextProps)
      }
    }

    render() {
      if (!this.staticSheet) return null

      const {children, className, ...attrs} = this.props

      const props = filterProps(attrs)
      const tagClass = composeClasses([
        this.staticSheet.classes[staticTagName],
        this.dynamicSheet.classes[this.dynamicTagName],
        className
      ])

      return createElement(tagName, {...props, className: tagClass}, children)
    }
  }
}

export default styled
