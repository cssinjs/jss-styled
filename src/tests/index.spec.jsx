import renderer from 'react-test-renderer'
import React from 'react'

import styled, {Styled} from '../'
import styledDOM, {Styled as StyledDOM} from '../dom'

import CreateAppStyled from './AppStyled'
import CreateAppStyledDOM from './AppStyledDOM'


describe('Default styled functions tests', () => {
  it('renders correctly App with default styled', () => {
    const App = CreateAppStyled(styled)
    const tree = renderer.create(<App />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly App with default Styled', () => {
    const customStyled = Styled({
      baseButton: {
        color: 'red',
      },
    })

    const App = CreateAppStyled(customStyled)
    const tree = renderer.create(<App />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('DOM styled functions tests', () => {
  it('renders correctly App with default styled', () => {
    const App = CreateAppStyledDOM(styledDOM)
    const tree = renderer.create(<App />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly App with default Styled', () => {
    const customStyled = StyledDOM({
      baseButton: {
        color: 'red',
      },
    })

    const App = CreateAppStyledDOM(customStyled)
    const tree = renderer.create(<App />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
