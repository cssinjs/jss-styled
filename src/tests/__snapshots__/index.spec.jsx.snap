exports[`test renders correctly App with default Styled 1`] = `
<div
  className="div-1-0-9">
  <header
    className="header-2-0-10">
    <h1
      className="h1-5-0-11">
      Title
    </h1>
  </header>
  <section
    className="section-3-0-12"
    data-name="content">
    <button
      className="button-6-0-13 button-11-0-14">
      primitive test
    </button>
    <button
      className="button-6-0-13 button-12-0-15">
      dynamic primitive test
    </button>
  </section>
  <section
    className="section-4-0-16">
    Another section
  </section>
</div>
`;

exports[`test renders correctly App with default styled 1`] = `
<div
  className="div-1-0-0">
  <header
    className="header-2-0-1">
    <h1
      className="h1-5-0-2">
      Title
    </h1>
  </header>
  <section
    className="section-3-0-3"
    data-name="content">
    <button
      className="button-6-0-4 button-11-0-5">
      primitive test
    </button>
    <button
      className="button-6-0-4 button-12-0-6">
      dynamic primitive test
    </button>
  </section>
  <section
    className="section-4-0-7">
    Another section
  </section>
</div>
`;

exports[`test renders correctly App with injectStyled 1`] = `
<div
  className="root-0-17">
  <div
    className="div-1-0-19">
    <header
      className="header-2-0-20">
      <h1
        className="h1-5-0-21">
        Title
      </h1>
    </header>
    <section
      className="section-3-0-22"
      data-name="content">
      <button
        className="button-6-0-23 button-11-0-24">
        primitive test
      </button>
      <button
        className="button-6-0-23 button-12-0-25">
        dynamic primitive test
      </button>
    </section>
    <section
      className="section-4-0-26">
      Another section
    </section>
  </div>
</div>
`;
